src/seq_futures.rs

```diff
@@ -0,0 +1,351 @@
+/// [`StreamExt::collect`]: futures::stream::StreamExt::collect
+/// [`StreamExt::buffered`]: futures::stream::StreamExt::buffered
+pub fn seq_join<S, Fut, O>(active: NonZeroUsize, stream: S) -> SequentialFutures<S, Fut, O>
+where
+    S: Stream<Item = Fut> + Send,
+    Fut: IntoFuture<Output = O> + ?Sized,
+{
+    SequentialFutures {
+        stream: stream.fuse(),
+        active: VecDeque::with_capacity(active.get()),
+    }
+}
+
+/// A substitute for [`futures::future::try_join_all`] that uses [`seq_join`].
+/// This awaits all the provided futures in order,
+/// aborting early if any future returns `Result::Err`.
+///
+/// [`seq_join`]: raw_ipa::helpers::buffers::seq_join
+pub async fn seq_try_join_all<I, O, E>(futures: I) -> Result<Vec<O>, E>
+where
+    I: IntoIterator,
+    I::IntoIter: Send,
+    I::Item: IntoFuture<Output = Result<O, E>>,
+{
+    const ACTIVE: Option<NonZeroUsize> = NonZeroUsize::new(256);
+
+    let futures = futures.into_iter();
+    let (lower, upper) = futures.size_hint();
+    let mut res = Vec::with_capacity(lower.saturating_add(upper.unwrap_or(0)));
+    let mut s = seq_join(ACTIVE.unwrap(), iter(futures));
+    while let Some(r) = s.next().await {
+        res.push(r?);
+    }
+    Ok(res)
+}
+
+/// A substitute for [`futures::future::join_all`] that uses [`seq_join`].
+///
+/// [`seq_join`]: raw_ipa::helpers::buffers::seq_join
+pub async fn seq_join_all<I, O>(futures: I) -> Vec<O>
+where
+    I: IntoIterator,
+    I::IntoIter: Send,
+    I::Item: IntoFuture<Output = O>,
+{
+    const ACTIVE: Option<NonZeroUsize> = NonZeroUsize::new(256);
+
+    seq_join(ACTIVE.unwrap(), iter(futures))
+        .collect::<Vec<_>>()
+        .await
+}
+
+#[pin_project]
+pub struct SequentialFutures<S, F, O>
+where
+    S: Stream<Item = F> + Send,
+    F: IntoFuture<Output = O> + ?Sized,
+{
+    #[pin]
+    stream: Fuse<S>,
+    active: VecDeque<Pin<Box<F::IntoFuture>>>,
+}
+
+impl<S, F, O> Stream for SequentialFutures<S, F, O>
+where
+    S: Stream<Item = F> + Send,
+    F: IntoFuture<Output = O>,
+{
+    type Item = O;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+
+        // Draw more values from the input, up to the capacity.
+        while this.active.len() < this.active.capacity() {
+            if let Poll::Ready(Some(f)) = this.stream.as_mut().poll_next(cx) {
+                this.active.push_back(Box::pin(f.into_future()));
+            } else {
+                break;
+            }
+        }
+
+        if let Some(f) = this.active.front_mut() {
+            if let Poll::Ready(v) = Future::poll(Pin::as_mut(f), cx) {
+                drop(this.active.pop_front());
+                Poll::Ready(Some(v))
+            } else {
+                for f in this.active.iter_mut().skip(1) {
+                    let res = Future::poll(Pin::as_mut(f), cx);
+                    assert!(res.is_pending(), "future resolved out of order");
```

Comment:
I think there may be a race that will cause a false positive here, if the first future has not resolved when it is checked on line 141, and then multiple futures resolve in quick succession.